File: BaselineSort.swift
import Foundation

func moveLeft(_ a: inout [Int], n: Int) {
    var pos = 0
    for i in 0..<n {
        if a[i] != 0 {
            a[pos] = a[i]
            pos += 1
        }
    }
    for i in pos..<n {
        a[i] = 0
    }
}

print("write your numbers")
let n = Int(readLine()!)!
let inputStr = readLine()!
let parts = inputStr.components(separatedBy: " ").filter{!$0.isEmpty}

var arr:[Int] = []
for s in parts {
    arr.append(Int(s)!)
}

var data = arr
var res = [Int]()
var baseSum = 0
var removed = Array(repeating: false, count: n) //标记线段是否取出

while res.count < n {
    // 找当前未移除、非0元素的最小高度
    var minH = 999999
    for i in 0..<n {
        if (!removed[i]) && data[i] != 0 && data[i] < minH {
            minH = data[i]
        }
    }
    
    baseSum += minH
    
    // 所有未移除的线段减去minH
    for i in 0..<n {
        if (!removed[i]) {
            data[i] -= minH
        }
    }
    
    // 收集本轮归零的线段
    for i in 0..<n {
        if (!removed[i] && data[i] == 0) {
            res.append(baseSum)
            removed[i] = true //标记这条线段已经拿走
        }
    }
    
    moveLeft(&data, n: n)
}

print("The result is: \(res)")
let resDesc = Array(res.reversed())
print("The descending result is: \(resDesc)")

File: BaselineSort.cpp
#include <iostream>
using namespace std;
void moveLeft(int a[], int n){
    int pos = 0;
    for(int i = 0; i < n; i++){
        if(a[i] != 0){
            a[pos] = a[i];
            pos++;
        }
    }
    for(int i = pos; i < n; i++)
        a[i] = 0;
}

int main()
{
    int arr[100];
    int n;
    cout << "输入数字个数：";
    cin >> n;
    for(int i = 0; i < n; i++)
        cin >> arr[i];

    int data[100];
    for(int i = 0; i < n; i++)
        data[i] = arr[i];

    int res[100] = {0};
    int resCnt = 0;
    int totalBase = 0; // 累计减去的总和！！

    for(int round = 0; round < n; round++)
    {
        int minv = 1000000000;
        for(int i = 0; i < n; i++)
        {
            if(data[i] != 0 && data[i] < minv)
            {
                minv = data[i];
            }
        }
        if(minv == 1000000000)
            break;

        totalBase += minv; // 叠加基线
        res[resCnt] = totalBase; // 累加后的才是原始数值！

        for(int i = 0; i < n; i++)
        {
            if(data[i] != 0)
                data[i] -= minv;
        }
        moveLeft(data, n);

        resCnt++;
    }

    cout << "升序结果：";
    for(int i = 0; i < resCnt; i++)
        cout << res[i] << " ";
    return 0;
}
